hypervisor/arch/x86/trusty.c

```diff
@@ -50,6 +50,222 @@ static struct key_info g_key_info = {
 	exec_vmwrite(VMX_GUEST_##SEG_NAME##_ATTR, seg.attr); \
 }
 
+#ifdef WORKAROUND_FOR_TRUSTY_4G_MEM
+
+uint64_t gpa2hpa_for_trusty(struct vm *vm, uint64_t gpa)
+{
+	uint64_t hpa = 0;
+	struct entry_params entry;
+	struct map_params map_params;
+
+	map_params.page_table_type = PTT_EPT;
+	map_params.pml4_base = HPA2HVA(vm->arch_vm.sworld_eptp);
+	map_params.pml4_inverted = HPA2HVA(vm->arch_vm.m2p);
+	obtain_last_page_table_entry(&map_params, &entry, (void *)gpa, true);
+	if (entry.entry_present == PT_PRESENT) {
+		hpa = ((entry.entry_val & (~(entry.page_size - 1)))
+				| (gpa & (entry.page_size - 1)));
+		pr_dbg("GPA2HPA: 0x%llx->0x%llx", gpa, hpa);
+	} else {
+		pr_err("VM %d GPA2HPA: failed for gpa 0x%llx",
+				vm->attr.boot_idx, gpa);
+	}
+
+	return hpa;
+}
+
+/**
+ * @defgroup trusty_apis Trusty APIs
+ *
+ * This is a special group that includes all APIs
+ * related to Trusty
+ *
+ * @{
+ */
+
+/**
+ * @brief Create Secure World EPT hierarchy
+ *
+ * Create Secure World EPT hierarchy, construct new PML4/PDPT,
+ * reuse PD/PT parse from vm->arch_vm->ept
+ *
+ * @param vm pointer to a VM with 2 Worlds
+ * @param gpa_orig original gpa allocated from vSBL
+ * @param size LK size (16M by default)
+ * @param gpa_rebased gpa rebased to offset xxx (511G_OFFSET)
+ *
+ */
+static void create_secure_world_ept(struct vm *vm, uint64_t gpa_orig,
+		int64_t size, uint64_t gpa_rebased)
+{
+	uint64_t nworld_pml4e = 0;
+	uint64_t sworld_pml4e = 0;
+	struct entry_params entry;
+	struct map_params map_params;
+	uint64_t gpa_uos = gpa_orig;
+	uint64_t gpa_sos;
+	uint64_t adjust_size;
+	uint64_t mod;
+	uint64_t hpa = gpa2hpa(vm, gpa_uos);
+	void *sub_table_addr = NULL, *pml4_base = NULL;
+	int i;
+	struct vm *vm0 = get_vm_from_vmid(0);
+	struct vcpu *vcpu;
+
+	if (vm0 == NULL) {
+		pr_err("Parse vm0 context failed.");
+		return;
+	}
+
+	if (!vm->sworld_control.sworld_enabled
+			|| vm->arch_vm.sworld_eptp != 0UL) {
+		pr_err("Sworld is not enabled or Sworld eptp is not NULL");
+		return;
+	}
+
+	/* Backup secure world info, will be used when
+	 * destroy secure world
+	 */
+	vm->sworld_control.sworld_memory.base_gpa = gpa_rebased;
+	vm->sworld_control.sworld_memory.base_hpa = hpa;
+	vm->sworld_control.sworld_memory.length = size;
```

Comment:
size is int64_t (is this expected?), while length is uint64_t. Need an explicit conversion here.